### ISSUE TITLE
Memory leak?

### ISSUE THREAD

**Issue by selkordy** [warpdotdev/Warp#5357]
### Dupe Check

- [X] I have [searched Warp bugs](https://github.com/warpdotdev/warp/issues?q=is%3Aissue+label%3ABugs) and there are no duplicates

### Describe the bug

Warp is using 100gb memory on my Mac. No doubt I've got a few tabs open but I only have a few remote SSH sessions going.  It's leaking somewhere. 

### To reproduce

Not sure, but I've got ~8 tabs open with maybe ~15 panes

from top
ID    COMMAND      %CPU TIME     #TH    #WQ  #PORT MEM    PURG  CMPRS PGRP  PPID  STATE
527    stable       5.3  04:09:56 73     3    461   101G+  512M  76G-  527   1     sleeping

### Operating system

MacOS

### Operating system and version

Sonoma 14.6.1 (23G93)

### Current Warp version

v0.2024.09.24.08.02.stable_01

### Regression

Yes, this bug started recently or with an X Warp version

### Recent working Warp date

Not sure, but I would have noticed a process using 100gb of memory.  I believe this was introduced in the latest release.

### Additional context

As of last week I did receive a few popups from my Mac that I was low on mem, but checking top nothing was using an excessive amount of memory and showed plenty of free memory available. I also received OOM errors in other apps.  I now believe Warp has a memory leak. 

### Does this block you from using Warp daily?

No

### Is this an issue only in Warp?

Yes, I confirmed that this only happens in Warp, not other terminals.

### Warp Internal (ignore): linear-label:b9d78064-c89e-4973-b153-5178a31ee54e

None

**Comment by vorporeal** [comment]
Hey @selkordy - thanks for letting us know!  Definitely something very wrong here.

I assume you've restarted Warp since then?  If so, have you experienced runaway memory usage again?

**Comment by selkordy** [comment]
Yes I've restarted since and haven't seen such excessive memory usage. 

I have a hunch the problem is related to tqdm. I had many tqdms going with leave=False running a process that took days to run accumulating memory.

**Comment by stanleyume** [comment]
Same here.
60GB memory usage on Mac
Two tabs. Two panes on one tab.

It builds up when you leave it open. I've been using it daily for a couple weeks.
So I usually just restart it periodically.

**Comment by demyxco** [comment]
Any news on this? I'm also having the same issue. Swap seems to be almost full too. Last screenshot was taken during a `cmd + q` on the app. Looks like memory is being freed up in real time.

![Screenshot 2024-12-02 at 14 12 15](https://github.com/user-attachments/assets/71dd2c8c-3a79-4235-8bd5-d4e711787efb)
![Screenshot 2024-12-02 at 14 07 17](https://github.com/user-attachments/assets/04b067f3-930e-4a46-ad42-bb137d6b1710)
![Screenshot 2024-12-02 at 14 13 57](https://github.com/user-attachments/assets/daf94cab-fc6b-4ef1-87ca-7604eb6f1ea4)

**Comment by vorporeal** [comment]
Hey all - we discovered recently that the macOS Activity Monitor will attribute memory usage of _all processes that are descendants of Warp_ to the Warp app.  (In general, it prefers to provide resource consumption metrics at an "app" level, not a process level.)

This means that if you're running a process in your shell (a tool or a program you're developing) that gobbles up memory, the Activity Monitor will show that as Warp using up a ton of memory.

When this next happens, can you take a look at a _process_ monitoring tool (e.g.: `top`/`htop`/`btop`) and see what the resident set size of the `warp` process is?  Would be good to know if Warp is consuming memory or it's actually some other process.

(Someone on our team ran into this issue, and it turned out that the `golangci-lint` process they were running in a shell session was using 30GB of RAM, and Warp was actually only using 500MB.)

**Comment by demyxco** [comment]
@vorporeal it happened again. I'm a sysadmin so I'm mainly using warp to ssh into several servers at any given time and usually leave it open. I don't have anything using `tail` or executing a binary that constantly streams to stdout so I'm not understanding what is causing this memory leak.
![Screenshot 2024-12-05 at 10 42 11](https://github.com/user-attachments/assets/0408ec13-6892-4795-b613-85cc53ffd6c8)
![Screenshot 2024-12-05 at 10 44 49](https://github.com/user-attachments/assets/9b061418-cf8f-480b-b3ae-fd0929187c11)
![Screenshot 2024-12-05 at 10 45 58](https://github.com/user-attachments/assets/bb0c3bc1-1744-4c4f-9cf0-5a277ef79cef)

**Comment by vorporeal** [comment]
Yeah, so "Real Memory Size" ("RES", in the `htop` image) is the thing that's most relevant here - Warp is only actually using 427MB of physical memory.  "Virtual Memory Size" represents how much memory is mapped in the virtual memory subsystem, but that also includes things like empty/zero pages, which don't actually map to any real/physical memory.

In the tree of processes under the `stable --sever` process (pid 63105), are there any with a large value in the "RES" column (i.e.: anything in the multiple gigabytes range)?

**Comment by demyxco** [comment]
I didn't think you'd reply so fast but unfortunately I already relaunched the app. I'll check when/if it happens again. You guys have any idea what's going on or a fix on the way?

**Comment by vorporeal** [comment]
My best understanding at the moment is it's an _attribution_ issue - Activity Monitor is attributing the memory consumption of a descendant process to Warp.app, but Warp itself isn't responsible for that memory usage.

**Comment by falrom** [comment]
I encountered a similar issue on my Ubuntu 22 PC. I found that after connecting to a Docker container, even without doing anything, Warp's memory usage increases by 2MB per second and never stops.

**Comment by falrom** [comment]
> I encountered a similar issue on my Ubuntu 22 PC. I found that after connecting to a Docker container, even without doing anything, Warp's memory usage increases by 2MB per second and never stops. If I disconnect from the Docker container, the memory usage stops increasing. However, the memory that has already been consumed is not released, even after I close the tab.

I did some more testing. Simply attaching to an already running Docker container does not, by itself, cause a memory leak.
However, once I execute the command 'Warpify subshell Ctrl+I', memory usage begins to grow endlessly.

**Comment by vorporeal** [comment]
@falrom I can reproduce this - thanks for the detailed report!  I'll take a look today and see if I can get a fix into this week's release.

**Comment by happyf-weallareeuropean** [comment]
same here.  macos started frqly pop me to quit app freeup ram i saw warp is 9gb ram everi time. wild is in the activity monitor just show 300mb. i dont know whats happen but something

**Comment by happyf-weallareeuropean** [comment]
> same here. macos started frqly pop me to quit app freeup ram i saw warp is 9gb ram everi time. wild is in the activity monitor just show 300mb. i dont know whats happen but something

i run some small backend server 24/7 with logs 1 tab.

**Comment by zh-impact** [comment]
I use both Windows and macOS machines, but the issue is more severe on Windows.

Even after closing all but one tab, the memory usage doesn't decrease.
<img width="1768" height="331" alt="Image" src="https://github.com/user-attachments/assets/1e4b2ab3-4ff5-49ef-b6ba-3bae9cafbedd" />

<img width="2148" height="765" alt="Image" src="https://github.com/user-attachments/assets/a687374c-8a62-4f7a-9f2f-46917ff137d5" />

<img width="1815" height="492" alt="Image" src="https://github.com/user-attachments/assets/ed36a2ed-d634-4211-b4f1-28076caa10f5" />

**Comment by vorporeal** [comment]
> [@falrom](https://github.com/falrom) I can reproduce this - thanks for the detailed report! I'll take a look today and see if I can get a fix into this week's release.

Forgot to follow up here at the time - this has been live for ~two weeks now.

---

Want to acknowledge the couple additional comments above.  I'm trying to come up with ideas for how we can better diagnose what's going on here; will follow up when I have more updates.

**Comment by Openliquid** [comment]
Same problem here on different macs, sometimes warp uses over 100gb memory.

**Comment by Aeromix** [comment]
This happens to me and ONLY with Warp credit use. I run out fast and end up using API Keys and it doesn't happen with API key use. I always now need to close Warp in between (Unreal) builds because it pegs my CPU at 100% (64 Core Threadripper/256GB RAM) and GRINDS my PC to a HALT. Takes a good 5 minutes to manage to get Warp to close. Insane memory leak to make a computer as powerful as mine crawl on its knees.

I love Warp but this is a major issue and wastes so much time if I dont catch it and it ends up just sitting there stalled for 30 minutes before I notice. No way thats good for hardware either.

This is on Windows, I've seen it on my Mac too. I do builds externally in terminal, so its not a sub process. But just now I saw last time it was pegged, Warp was running 1113 processes. What could that possibly be when its idle and not in use?

**Comment by gywhitel** [comment]
I also found this problem, which seems related to warpified ssh terminal. Warp takes me 51.7 GB memory when a ssh warp running. But it goes back normal when I close the ssh terminal.

<img width="1652" height="134" alt="Image" src="https://github.com/user-attachments/assets/cfc57acb-5d2d-488c-b73a-40c19b2197ed" />

**Comment by kevinchevalier** [comment]
Ah, interesting. Does this happen pretty quickly or does it take a long time to build up? Minutes vs hours?

What are you doing in the SSH session? Running a long running command with lots of output, or something else?

**Comment by kevinchevalier** [comment]
For anyone following, we have shipped a few memory usage improvements over the past month, and will ship more today. This is my highest priority at the moment.

Try out a new version if it's been a while, and let me know if you notice memory issues. The Preview version also has automatic heap reporting (on macOS) if using more than 10GB.

https://www.warp.dev/download-preview

**Comment by bandesz** [comment]
I was running a large C build a couple of times with large amount of logs. Warp killed my MacBook with 64Gb memory, reporting that it was consuming 96Gb. I don't think it was wise to ask Warp to run the build commands 😄

**Comment by kevinchevalier** [comment]
Hey @bandesz Can you provide a few more details? What operating system? 

I would love to explore if we can debug this issue for you.

**Comment by bandesz** [comment]
@kevinchevalier Tahoe 26.4.1. Warp version was either v0.2026.04.29.08.56.stable_00 or v0.2026.05.06.15.42.stable_05.

I just tried to replicate it with v0.2026.05.06.15.42.stable_05 again, but I couldn't this time.

**Comment by kevinchevalier** [comment]
Ok, was it pretty reproducible before? Please let me know if you notice it again. I'm focused on squashing any memory issues that we can find in Warp.
